aws/resource_aws_lambda_layer.go

```diff
@@ -0,0 +1,262 @@
+package aws
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+
+	arn2 "github.com/aws/aws-sdk-go/aws/arn"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+
+	"github.com/aws/aws-sdk-go/aws"
+
+	"github.com/aws/aws-sdk-go/service/lambda"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+const awsMutexLambdaLayerKey = `aws_lambda_layer`
+
+func resourceAwsLambdaLayer() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsLambdaLayerPublish,
+		Read:   resourceAwsLambdaLayerRead,
+		Delete: resourceAwsLambdaLayerDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"layer_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"filename": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"s3_bucket", "s3_key", "s3_object_version"},
+			},
+			"s3_bucket": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"filename"},
+			},
+			"s3_key": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"filename"},
+			},
+			"s3_object_version": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"filename"},
+			},
+			"compatible_runtimes": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				ForceNew: true,
+				MinItems: 0,
+				MaxItems: 5,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validation.StringInSlice(validLambdaRuntimes, false),
+				},
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"license_info": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(0, 512),
+			},
+
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"qualified_arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"created_date": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"source_code_hash": {
```

Comment:
We should consider adding an acceptance test utilizing `create_before_destroy` and this attribute to verify that Lambda is happy with this type of replacement workflow. e.g. in the first configuration

```hcl
resource "aws_lambda_layer_version" "test" {
  filename         = "test-fixtures/lambdatest.zip"
  layer_name       = %q
  source_code_hash = "${base64sha256(file("test-fixtures/lambdatest.zip"))}"

  lifecycle {
    create_before_destroy = true
  }
}
```

Then updating which test fixture zip file its pointing at in the second configuration.